### PR TITLE
Quick action widget

### PR DESCRIPTION
**Pull request by deeppandya**
Resolves https://github.com/brave/brave-browser/issues/22875
Resolves https://github.com/brave/internal/issues/942

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Please follow spec from :https://github.com/brave/internal/issues/942

### CODE REVIEW

**Comment by deeppandya** [comment]
Need to remove gain metadata files as it's not needed.

File: patches/chrome-android-java-src-org-chromium-chrome-browser-quickactionsearchwidget-QuickActionSearchWidgetProvider.java.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+diff --git a/chrome/android/java/src/org/chromium/chrome/browser/quickactionsearchwidget/QuickActionSearchWidgetProvider.java b/chrome/android/java/src/org/chromium/chrome/browser/quickactionsearchwidget/QuickActionSearchWidgetProvider.java
+index 171242579ac25e0f845d44f3bbd9c501e6e35d95..0c143fd08c1b7c94186dfef6f5dfff199f2e5b00 100644
+--- a/chrome/android/java/src/org/chromium/chrome/browser/quickactionsearchwidget/QuickActionSearchWidgetProvider.java
++++ b/chrome/android/java/src/org/chromium/chrome/browser/quickactionsearchwidget/QuickActionSearchWidgetProvider.java
+@@ -213,7 +213,7 @@ public abstract class QuickActionSearchWidgetProvider extends AppWidgetProvider
+             // Make the Widget available to all Chrome users who participated in an experiment in
+             // the past. This can trigger disk access. Unfortunately, we need to keep it for a
+             // little bit longer -- see: https://crbug.com/1309116
+-            setWidgetEnabled(true, true);
++            setWidgetEnabled(false, false);
```
Comment (samartnik) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-quickactionsearchwidget-QuickActionSearchWidgetProvider.java.patch:10]: I believe we can change method owner for `setWidgetEnabled` and replace values there.

File: third_party/gson/BUILD.gn [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2016 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import("//build/config/android/rules.gni")
+
+java_prebuilt("gson_java") {
```
Comment (samartnik) [inline third_party/gson/BUILD.gn:7]: Not sure how justified adding this lib for such a small class. It seems to me that adding own `toJson` and `fromJson` are more preferable. At least Chromium team goes this way.

File: patches/chrome-android-java-AndroidManifest.xml.patch [hunk 3]
```diff
@@ -79,15 +79,31 @@ index a3b0ac4f938183817cbd294d219c7294a4ff60e1..d36fd43f4ddea20bfe695277923f403e
              <meta-data android:name="android.app.searchable"
                  android:resource="@xml/searchable" />
          </activity-alias>
-@@ -1188,6 +1203,7 @@ by a child template that "extends" this file.
+@@ -891,6 +906,7 @@ by a child template that "extends" this file.
+         <receiver
+             android:name="org.chromium.chrome.browser.quickactionsearchwidget.QuickActionSearchWidgetProvider$QuickActionSearchWidgetProviderSearch"
+             android:label="@string/quick_action_search_widget_title"
++            android:enabled="false"
```
Comment (samartnik) [inline patches/chrome-android-java-AndroidManifest.xml.patch:86]: Don't we disable it in a runtime in `setWidgetEnabled`?

File: android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/QuickActionSearchAndBookmarkWidgetProvider.java [hunk 4]
```diff
@@ -0,0 +1,436 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/QuickActionSearchAndBookmarkWidgetProvider.java:1]: nit: 2022

File: android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/promo/SearchWidgetPromoPanel.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
+/*
+  Copyright (c) 2021 The Brave Authors. All rights reserved.
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/promo/SearchWidgetPromoPanel.java:2]: nit: 2022

File: android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/utils/BraveSearchWidgetUtils.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
+/*
+  Copyright (c) 2021 The Brave Authors. All rights reserved.
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/utils/BraveSearchWidgetUtils.java:2]: nit: 2022

File: android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/utils/BraveSearchWidgetUtils.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                pinnedWidgetCallbackIntent.putExtras(bundle);
+                PendingIntent successCallback =
+                        PendingIntent.getBroadcast(context, 0, pinnedWidgetCallbackIntent,
+                                PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT);
+
+                appWidgetManager.requestPinAppWidget(appWidgetProvider, null, successCallback);
+            }
+        }
+    }
+}
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/utils/BraveSearchWidgetUtils.java:64]: nit: last empty line

File: android/java/res/layout/layout_search_widget_promo.xml [hunk 8]
```diff
@@ -0,0 +1,110 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (c) 2020 The Brave Authors. All rights reserved.
```
Comment (samartnik) [inline android/java/res/layout/layout_search_widget_promo.xml:2]: nit: 2022

File: android/java/res/layout/quick_action_search_and_bookmark_widget_layout.xml [hunk 9]
```diff
@@ -0,0 +1,587 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--* Copyright (c) 2020 The Brave Authors. All rights reserved.
```
Comment (samartnik) [inline android/java/res/layout/quick_action_search_and_bookmark_widget_layout.xml:2]: nit: 2022

File: android/java/res/layout/layout_search_widget_promo.xml [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+                android:text="@string/not_now"
+                android:fontFamily="@font/poppins_medium"
+                android:textColor="@color/search_widget_promo_not_now_text_color"
+                app:layout_constraintBottom_toBottomOf="@id/cvAddWidget"
+                app:layout_constraintEnd_toEndOf="parent"
+                app:layout_constraintStart_toEndOf="@id/cvAddWidget"
+                app:layout_constraintTop_toTopOf="@id/cvAddWidget" />
+        </androidx.constraintlayout.widget.ConstraintLayout>
+    </androidx.cardview.widget.CardView>
+</FrameLayout>
```
Comment (samartnik) [inline android/java/res/layout/layout_search_widget_promo.xml:110]: nit: last empty line

**Review by samartnik — CHANGES_REQUESTED** [review]

File: android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/BraveQuickActionSearchWidgetProvider.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
+// Copyright 2022 The Chromium Authors. All rights reserved.
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/BraveQuickActionSearchWidgetProvider.java:1]: Please use our header with Mozilla v2 lincense

File: android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/QuickActionSearchAndBookmarkWidgetProvider.java [hunk 12]
```diff
@@ -0,0 +1,465 @@
...
+
+    private static PendingIntent createPendingIntent(Context context, Intent intent) {
+        return PendingIntent.getActivity(context, 0, intent,
+                PendingIntent.FLAG_UPDATE_CURRENT
+                        | IntentUtils.getPendingIntentMutabilityFlag(false));
+    }
+
+    /**
+     * This class acts as a single source of truth for this widet. This widget would use this class
+     *to fetch the data. Also, any modification to the data should be done through this class.
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/QuickActionSearchAndBookmarkWidgetProvider.java:358]: nit: space

File: android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/promo/SearchWidgetPromoPanel.java (hunk 5 above)
Comment (samartnik) [reply android/java/org/chromium/chrome/browser/widget/quickactionsearchandbookmark/promo/SearchWidgetPromoPanel.java:2]: It's still 2021 :)

File: android/java/res/layout/bookmark_widget.xml [hunk 13]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2015 The Chromium Authors. All rights reserved.
```
Comment (samartnik) [inline android/java/res/layout/bookmark_widget.xml:2]: We should use our header

File: android/java/res/layout/bookmark_widget_item.xml [hunk 14]
```diff
@@ -0,0 +1,47 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2022 The Chromium Authors. All rights reserved.
```
Comment (samartnik) [inline android/java/res/layout/bookmark_widget_item.xml:2]: Should be our header

**Review by samartnik — CHANGES_REQUESTED** [review]

**Review by samartnik — APPROVED** [review]
lgtm

**Review by mkarolin — APPROVED** [review]
`patches` ++

**Comment by AndyAnds265** [comment]
@deeppandya, any chance you could add a screenshot to this string?
        try_brave_search_widget

**Pull request merged**